Port line-sort-key and fill-sort-key

ahk opened the pull request:
Adds support for `line-sort-key` and `fill-sort-key` properties, consistent with `symbol-sort-key`.

It took me quite a while to gain proficiency with the C++ usage in the project (I was rustier than I thought). Consider this a draft, and I would love to hear your comments and suggestions. I'm sure I may have style or design issues with my approach. For quite a while I pursued a design that modified the `PatternLayout` contructor only (rather than introducing my new function `maybeApplySortKey`). This was difficult for me to figure out because fill extrusion layer type does not support `*-sort-key` but there is a decent chunk of shared code among fill, line, and fill extrusion layers around `Layout` creation (which is most of what `PatternLayout` class is). Hopefully someone can point me to a better design.

I attempted to be consistent with the original PR for `symbol-sort-key` as much as possible.

asheemmamoowala commented:
> I haven't tried to rebase this, it's obviously out of date. Is it preferred to rebase this against master or to merge master?

Rebasing is preferable. I prefer rebasing regularly so as not to fall behind too much.

ahk commented:
> > I haven't tried to rebase this, it's obviously out of date. Is it preferred to rebase this against master or to merge master?
> 
> Rebasing is preferable. I prefer rebasing regularly so as not to fall behind too much.

Rebased against master now, apparently that made the darwin platform code generation a noop.

pozdnyakov commented on src/mbgl/layermanager/fill_layer_factory.cpp:
```diff
@@ -29,8 +29,13 @@ FillLayerFactory::createLayout(const LayoutParameters& parameters,
+    using LayoutType = PatternLayout<FillBucket, FillLayerProperties, FillPattern, FillLayoutProperties>;
+    auto&& layout = std::make_unique<LayoutType>(parameters.bucketParameters, group, std::move(layer), parameters.imageDependencies);
+
+    layout->template maybeApplySortKey<FillSortKey>();
```
why call it here? could it be done inside `Layout` constructor?

ahk replied:
I wanted to do this originally, but we don't have `constexpr if` so I can't have the conditional `SortKeyType` template param useable in the constructor without breaking type safety (`FillExtrusion` shares this constructor and and has no `SortKey` property.) I was thinking about making a `PatternLayoutSorted` subclass ...

pozdnyakov replied:
e.g.
```
template <typename SortKeyPropertyType>
float calculateSortKey(const PatternFeature& feature)  {
.....
}
template <>
float calculateSortKey<void>(const PatternFeature&)  {
 return 0.0f;
}
````

ahk replied:
This is a great solution thank you, I had thought about using `void` but didn't think to make a specialization on it. Thanks for helping me with this, very useful knowledge!

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -18,15 +18,20 @@ using PatternLayerMap = std::map<std::string, PatternDependency>;
 class PatternFeature  {
 public:
-    const uint32_t i;
+    friend bool operator < (const PatternFeature& lhs, const PatternFeature& rhs) {
```
nit: const method

ahk replied:
"Non-member function cannot have 'const' qualifier"
Which I think is valid, and lhs and rhs are both const.

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -90,12 +96,27 @@ class PatternLayout : public Layout {
             }
-            features.push_back({static_cast<uint32_t>(i), std::move(feature), patternDependencyMap});
+
+            features.push_back({static_cast<uint32_t>(i), std::move(feature), patternDependencyMap, 0.});
```
could we evaluate the sort key already here?

pozdnyakov replied:
in that case we could keep `features` always sorted, the easiest way using of `std::set<PatternFeature>` instead of `std::vector`

ahk replied:
Maybe I don't know enough about `std::set`, but "being less than" seems to be the requirement for being unique. Because two separate `PatternFeature` may share a sort-key, wouldn't we end up with only one of them recorded in the `std::set`? I would need to come up with some other field to add to the comparison that would ensure all features are seen as "unique"?

ahk replied:
Also curious, what's the desire to do incremental update of the sort? This is so there's no intermediate state where the features are unsorted?

pozdnyakov replied:
If sort keys are not unique, we could use `std::list` + `std::lower_bound`

> Also curious, what's the desire to do incremental update of the sort? This is so there's no intermediate state where the features are unsorted?

Composing sorted list from the beginning is faster than compose it unsorted and then sort

ahk replied:
Huh, I didn't know that. I thought they were equivalent complexity and sorting in a batch was usually faster in practice because of reduced relocations and cache efficiency.

I thought about using `std::list` instead of `std::vec` so the features could be inserted incrementally without triggering relocations, but my memory was that `vec` perf in time and memory tends to be much better so I left it and used the batch sort. Happy to implement this with `std::list` though.

pozdnyakov commented:
The new approach looks great! @ahk what is the status of this pr? is it ready for review & merge (I see it is still a draft)?

ahk commented:
> is it ready for review & merge (I see it is still a draft)?

I still need to update the sorting logic to incrementally maintain the sort in the destination as you requested earlier. I'll clean that up and move it out of draft. It looks like it I'm failing more of the circleci ARM and Linux builds than I was before ... I might need some help figuring that out.

ahk commented:
Hi @pozdnyakov I've updated this branch to maintain the sort of features during layout step and use a std::list over a std::vector. I also found an indexing numeric limits bug.

I still can't figure out how to make the sanity-check format step happy. When I apply the patch it suggests and it reruns it shows me a new patch that undoes the previous patch, can't explain that one.

Also apparently some of our linux targets use a compiler that doesn't allow fully specializing a template function in a class template. I know partial specialization isn't allowed but if you have any tips on how to resolve that I would appreciate it. Is this a funny parsing quirk I'm not familiar with?

Moved this to review because I think it does everything it's supposed to and I resolved all the comments as requested.

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -19,15 +20,37 @@ using PatternLayerMap = std::map<std::string, PatternDependency>;
+                   const float sortKey_)
+        : i(i_), feature(std::move(feature_)), patterns(patterns_), sortKey(sortKey_) {}
+
+    PatternFeature(PatternFeature&& pt) noexcept
```
= default

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -19,15 +20,37 @@ using PatternLayerMap = std::map<std::string, PatternDependency>;
+    PatternFeature(PatternFeature&& pt) noexcept
+        : i(pt.i), feature(std::move(pt.feature)), patterns(pt.patterns), sortKey(pt.sortKey) {}
+
+    PatternFeature& operator=(PatternFeature&& pt) noexcept {
```
= default

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -20,26 +20,53 @@ using PatternLayerMap = std::map<std::string, PatternDependency>;
 
+template <class LayoutPropertiesType>
+struct PatternFeatureInserterUnsorted {
+    void insert(std::vector<PatternFeature>& features,
```
better make it static

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -20,26 +20,53 @@ using PatternLayerMap = std::map<std::string, PatternDependency>;
     float sortKey;
 };
 
+template <class LayoutPropertiesType>
```
instead, we could make a template method

pozdnyakov commented on src/mbgl/layout/pattern_layout.hpp:
```diff
@@ -163,23 +183,23 @@ template <class BucketType,
-    : public PatternLayout<BucketType, LayerPropertiesType, PatternPropertyType, LayoutPropertiesType> {
+          class LayoutPropertiesType,
+          class FeatureInserterType = PatternFeatureInserterSorted<LayoutPropertiesType, SortKeyPropertyType>>
+class PatternLayoutSorted : public PatternLayout<BucketType,
```
Looks like `PatternLayoutSorted` is not needed

chloekraw commented:
@alexshalamov thanks for taking this over the finish line!

1ec5 commented on scripts/generate-style-code.js:
```diff
@@ -8,8 +8,6 @@ const colorParser = require('csscolorparser');
 // FIXME: https://github.com/mapbox/mapbox-gl-native/issues/15008
 delete spec.layout_circle["circle-sort-key"]
-delete spec.layout_line["line-sort-key"]
-delete spec.layout_fill["fill-sort-key"]
```
These lines are duplicated in the Darwin version of this script:

https://github.com/mapbox/mapbox-gl-native/blob/1f266b45bc1cf5bec43cbd81a2da587023124401/platform/darwin/scripts/generate-style-code.js#L20-L21

1ec5 replied:
mapbox/mapbox-gl-native-ios#179 updates the gl-native-ios repository to reflect this change. @tobrun, heads-up that the Android map SDK is missing runtime styling methods due to similar delete statements in platform/android/scripts/generate-style-code.js.

tobrun replied:
@Chaoba could you follow up on above?
